compiler/qsc_eval/src/lower.rs

```diff
@@ -175,30 +196,23 @@ impl Lowerer {
             input,
             output,
             functors,
-            body,
-            adj,
-            ctl,
-            ctl_adj,
+            implementation,
         }
     }
 
     fn lower_spec_decl(&mut self, decl: &hir::SpecDecl) -> fir::SpecDecl {
+        let (controls, block) = if let SpecBody::Impl(pat, block) = &decl.body {
+            let pat_id = pat.as_ref().map(|p| self.lower_spec_decl_pat(p));
+            let block_id = self.lower_block(block);
+            (pat_id, block_id)
+        } else {
+            panic!("if a SpecDecl is some, then it must be an implementation");
+        };
```

Comment:
You can simplify this one into a let-else pattern:

```suggestion
        let SpecBody::Impl(pat, block) = &decl.body else {
            panic!("if a SpecDecl is some, then it must be an implementation");
        };
        let controls = pat.as_ref().map(|p| self.lower_spec_decl_pat(p));
        let block = self.lower_block(block);
```

Reply:
This is very neat! Updated.